src/foam/nanos/menu/Menu.js

```diff
@@ -70,13 +82,77 @@ foam.CLASS({
       name: 'icon',
       documentation: 'Icon associated to the menu item.',
       displayWidth: 80
+    },
+    {
+      documentation: 'Predicate providing arbitrary checks, in addition to the regular menu auth checks.',
+      class: 'foam.mlang.predicate.PredicateProperty',
+      name: 'readPredicate',
+      view: {
+        class: 'foam.u2.view.JSONTextView'
+      },
+      javaFactory: `
+      return foam.mlang.MLang.TRUE;
+      `,
     }
   ],
 
   methods: [
     function launch_(X, e) {
       this.lastMenuLaunchedListener && this.lastMenuLaunchedListener(this);
       this.handler && this.handler.launch(X, this, e);
+    },
+    {
+      documentation: 'Desire to call read predicate with calling context but predicate may also need access to this menu; add the current menu as context key MENU',
+      name: 'f',
+      type: 'Boolean',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaCode: `
+      return getReadPredicate().f(
+        x.put("MENU", this)
+      );
+      `
+    },
+    {
+      name: 'authorizeOnCreate',
+      javaCode: `
+      AuthService auth = (AuthService) x.get("auth");
+      if ( ! auth.check(x, "menu.create") ) {
+        throw new AuthorizationException("You do not have permission to create menus.");
+      }
+      `
+    },
+    {
+      name: 'authorizeOnUpdate',
+      javaCode: `
+      AuthService auth = (AuthService) x.get("auth");
+      if ( ! auth.check(x, "menu.update."+getId()) ) {
+        throw new AuthorizationException("You do not have permission to update this menu.");
+      }
+      `
+    },
+    {
+      name: 'authorizeOnDelete',
+      javaCode: `
+      AuthService auth = (AuthService) x.get("auth");
+      if ( ! auth.check(x, "menu.delete") ) {
```

Comment:
To stay consistent with the rest of the system, this should be "remove", not "delete" and should also include the id at the end.

```suggestion
      if ( ! auth.check(x, "menu.remove." + getId()) ) {
```